cython version of the algorithm

ev-br opened the pull request:
Here is a Cython version of the 3D skeletonization algorithm. The implementation is close to ready from my end, modulo several warts I'll comment on below.

The python entry point is `skel.py`, the actual implementation is in `_skel.pyx`.

So far it's not packaged anyhow, so the build/test sequence is
```
$ python setup.py build_ext --inplace
$ cp skeletonize_3d/_skel.so .
$ nosetests test_simple_2d.py
```

A couple of questions of the "how is this done in scikit-image" sort:

- What is the usual way to add documentation -- is it pulled from the docstrings, special notebooks, something else?
- I'm using as a test case the Bat Cochlea image from FIJI samples. This is the image from the front page of http://fiji.sc/Skeletonize3D. Not sure if there is some license issue.
- Thinning can be non-instant for larger size images. Is there / what is a recommended way for logging / progress reporting? 
- I assume `np.uint8` is an adequate dtype for a binary image, or is it?
- How strict we want to be if the user input is not of a compatible dtype? As a data point, Fiji refuses to process the image unless it's 8-bit monochrome. My gut feeling is to leave conversions to the user.
- The commit history is way too atomic. I kept it that way so far in case I'd need bisecting :-).

ev-br commented:
Now, on to @jni's comments on the early draft, https://github.com/scikit-image/skeleton_3D/pull/2#issuecomment-171147524:

> Because NumPy uses C-order arrays, we use p, r, c for the coordinates, see coordinate conventions

Done :-).

> get_neighborhood: this can be done much more readably and concisely using numpy indexing: return img[p-1:p+1, r-1:r+1, c-1:c+1]. In fact, maybe you don't need a function at all... =)

I'd rather not, for two reasons. First, the ordering of the pixels in the neighborhood here corresponds to both the Java code *and* the original paper, [Lee94]. Changing the order of pixels would require changing the numbering throughout. This is error-prone and I'd argue is *less* maintainable than the code with indexing which can be immediately checked against the paper.

Second, this is one of those small functions in the hot loop (tm), so it *really* pays off to squeeze the last drop of performance from it. From profiling, we're talking a factor of 3-5 of the total runtime as compared to making `neighbors` a C array (the current implementation).

> the LUT is inherently messy, but a cleaner alternative would be LUT[1::2] = [1, -1, -1, 1, -3, ..., 1, 1, -1] (with this list split over several lines). Is there any rhyme or reason to the grouping of values?

The layout is a direct retranslation from Java, adjusted for the syntax :-). More importantly, this LUT is a retype of a table 3 from [Lee94], and the grouping of values corresponds to the columns of the table.

With the current layout, checking a value against the paper or Java implementation is immediate. 
I can surely reformat it if you want me to --- I guess it's the question to maintainers. 

> octree_labeling is very messy and hard to maintain. I would suggest abstracting away the logic (which is the same in every case) from the values, which change. e.g.

Yeah, it's a mess :-(. I implemented a version similar to Juan's suggestion in https://github.com/ev-br/skeleton_3D/tree/octree_dynamic (the relevant commit is 90b95b95c4317aa19215c11a98f57a62f36995b4). The result is much more concise, and 3-5 times slower :-).

A viable alternative I see is to use build-time templating with Temptita which is bundled with cython anyway. I'll implement this in the coming days.

stefanv commented on _skel.pyx:
```diff
@@ -0,0 +1,748 @@
+    """
+    LUT = np.zeros(256, dtype=np.intc)
+
+    LUT[1]  =  1
```
Maybe we can find a better way of filling out this LUT!  What about:

```np.array([1, -1, -1, 1, -3...```

stefanv commented on _skel.pyx:
```diff
@@ -0,0 +1,748 @@
+
+    See index_octants and is_Euler_invariant routines below.
+    """
+    NEB, NWB, SEB, SWB, NEU, NWU, SEU, SWU = tuple(range(8))
```
No need to convert to tuple

stefanv commented on _skel.pyx:
```diff
@@ -0,0 +1,748 @@
+    See index_octants and is_Euler_invariant routines below.
+    """
+    NEB, NWB, SEB, SWB, NEU, NWU, SEU, SWU = tuple(range(8))
+    _neighb_idx = np.empty((8, 7), dtype=np.intc)
```
Again, we can probably just fill up the array directly:

```
np.array([[2, 1, 11, 10, 5, 4, 14],
          [0, 9, 3, 12, 1, 10, 4]...
```

stefanv commented on _skel.pyx:
```diff
@@ -0,0 +1,748 @@
+    # copy neighbors for labeling
+    # ignore center pixel (i=13) when counting (see [Lee94])
+    cdef pixel_type cube[26]
+    memcpy(cube, neighbors, 13*sizeof(pixel_type))
```
Can we get away without doing memcopies?  Arrays/memoryviews support .copy

ev-br replied:
Not easily. Using memoryviews here, let alone numpy arrays, is a major bottleneck (cf https://github.com/ev-br/skeleton_3D/commit/56999753a8a939cb7a6eff2417f05646941e0a02, https://github.com/ev-br/skeleton_3D/commit/5f3d87dbaba81d253f7e95f75a0ead52e9b92b48). One might get away with avoiding copying altogether, but that would throw off all indexing in `octree_labeling`, so I'd like to keep this as is until we finish up with the latter.

stefanv commented on _skel.pyx:
```diff
@@ -0,0 +1,748 @@
+    components in the 3D neighborhood after the center pixel would
+    have been removed.
+
+    See Figs. 6 and 7 of [Lee94] for the values of indices.
```
References (everywhere) should be of form ``[Lee94]_``.

stefanv commented:
Re: dataset

```
bat-cochlea-volume.zip
----------------------
This is a thresholded stack of the scala media volume within the
cochlea of the moustached bat.  The scale is 40 pixels per millimeter.
The original data set is from a uniform serial section of high
reslolution MRI done at the Duke Center for inVivo Microscopy
with tissue from the Dept. of Cell Biology and Anatomy at the
University of North Carolina.

Art Keating (keatinga@med.unc.edu)

To create a movie from this volume, first threshold the stack using
Image/Adjust/Threshold then perform Nearest Point projection using
Image/Stacks/Render.

Wayne Rasband (wayne@codon.nih.gov)
```

stefanv commented:
- I will mail the data author for permissions
- We generate the docs from docstrings
- ``np.uint8`` is what we use for binary images
- in scikit-image, we support most input types (see http://scikit-image.org/docs/stable/user_guide/data_types.html), but we have utility functions for converting to float, e.g., which is what most functions use internally.  You can choose what to use inside, and call another conversion function, ``img_as_xxx``

Thank you, this is already looking great!  Perhaps we can start thinking of making a WIP PR against scikit-image itself?  I thought this was going to take much longer, hence this repository.

emmanuelle commented:
Awesome! I had fun playing with the function, the result is really nice!
![skel_mayavi](https://cloud.githubusercontent.com/assets/263366/12695616/73b4dea0-c754-11e5-8919-79bbb0c6b3da.png)

emmanuelle commented:
@ev-br thank you so much for the beautiful work!

A few comments following on your questions:
- as for documentation, we follow NumPy's documentation standard to provide thorough and informative docstrings. This rule applies mostly to functions exposed in the API, but we try to have at least a "Parameters" section in all functions, including internal helper functions. In addition to docstring, we write example scripts that are converted to html and notebooks for the example gallery. See for example http://scikit-image.org/docs/dev/auto_examples/edges/plot_medial_transform.html#example-edges-plot-medial-transform-py
- for test cases and demos, you can also use the ``data.binary_blobs`` functions, that generate binary images with "blobs" (in 2D or 3D).

emmanuelle commented on skel.py:
```diff
@@ -0,0 +1,51 @@
+    """
+    img_oo = img_o[1:-1, 1:-1, 1:-1]
+    img_oo = img_oo.squeeze()
+    img_oo *= 255
```
if the original image has values in {0, 1}, I'm not very comfortable with values in {0, 255} for the output.

ev-br replied:
`img_oo *= img_in.max()` --- is this what you mean?

ev-br commented:
Thank you @stefanv, @emmanuelle for the review! I believe I addressed the review comments.

I also addressed @jni's request to separate the logic from boilerplate in `octree_labeling` in  f9001e3. This comes at the cost of slightly complicating the build process: 9efdf04 implements a slight tweak to  `skimage/_build.py`, with the difference adapted almost literally from https://github.com/scipy/scipy/blob/master/tools/cythonize.py.

As a data point, this sort of build-time templating causes no problems in scipy that I know of, but this is I guess up to you guys if you want this in scikit-image. If this is something we want, then I'll format this as a separate PR to the main scikit-image repo.

ev-br commented:
On the topic of directing this work as a PR to the main scikit-image repository, I'd appreciate suggestions for the user-facing API. Do we want it to be,
* a separate function in `skimage.morphology`, next to 2D `medial_axis` and `skeletonize` (where the latter implements, as far as I can see, similar but slightly different algorithm)
* a kwarg switch in `morphology.skeletonize`?
* just dispatch to this algorithm in `morphology.skeletonize` if the input image is 3D?

The latter option seems to minimize the API surface. OTOH, it might be useful to expose different algorithms for the 2D case, too.

stefanv commented:
Since this is a different algorithm, I think we should give it a new name.
We could call it `skeletonize_3d` or `skeletonize_name_of_algorithm` (not
sure what that is, or who published it).  A good home for this would then
be the ``morphology`` submodule.

Thanks again, this looks great.

ev-br commented:
Continued over at https://github.com/scikit-image/scikit-image/pull/1923

ev-br commented:
@emmanuelle what tool do you use to generate this image?
